# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Sidd13789/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I choose the **Content Refresh** lane because I want to learn how machine learning can help identify website pages that should be updated. This problem is practical and matches the dataset provided in this internship. I believe working on this lane will improve my machine learning skills while solving a real-world business problem.


# [link text](*https://*)## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [14]:
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
# Basic information
print("Dataset Shape:", df.shape)
print("\nColumns:\n", df.columns.tolist())

Dataset Shape: (30000, 44)

Columns:
 ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


**Research Question:** Which website pages should be refreshed first to prevent a decline in their performance?
**Decision:** Help the content or SEO team decide which pages should be prioritized for content updates.
**Action:** Refresh the highest-priority pages before their performance declines further.
**Cost of a Wrong Recommendation:** If the recommendation is wrong, important pages may be ignored while time and effort are spent updating pages that do not actually need attention. This can reduce website performance and waste resources.
**Why Machine Learning?** A website can contain thousands of pages, making manual prioritization difficult. Machine learning can analyze historical patterns and help rank pages that are more likely to require a content refresh, supporting faster and more consistent decision-making.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

 4. Careful words: what I can and can't claim`

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

This project is intended to support decision-making by identifying website pages that may need a content refresh based on patterns observed in the available data. The results should be interpreted as observations and recommendations, not as proof that updating a page will definitely improve its performance.This work cannot prove cause and effect or predict Google's ranking algorithm. It is designed to provide data-driven insights that help content and SEO teams prioritize their work more effectively, while recognizing that final decisions should also consider human expertise and business context.


In [15]:
!git clone https://github.com/Sidd13789/flyrank-ml-internship.git

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 127, done.
remote: Counting objects: 100% (127/127), done.
remote: Compressing objects: 100% (82/82), done.
remote: Total 127 (delta 42), reused 101 (delta 29), pack-reused 0 (from 0)
Receiving objects: 100% (127/127), 1.84 MiB | 16.38 MiB/s, done.
Resolving deltas: 100% (42/42), done.


In [16]:
%cd flyrank-ml-internship

/content/flyrank-ml-internship/flyrank-ml-internship


In [17]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print(df.shape)
print(df.head())

(30000, 44)
             content_id          client_id  search_volume  competition  \
0  content_304f48230142  client_f369cb89fc           10.0         0.67   
1  content_a1fb4e703a9e  client_4e07408562           90.0         0.01   
2  content_9aa793d4d895  client_7f2253d7e2            0.0         0.00   
3  content_331d6c4de07b  client_19581e27de           10.0         0.00   
4  content_d99b7a2d90ca  client_3fdba35f04            0.0         0.00   

  competition_level   cpc     content_type    main_intent  word_count  \
0              HIGH  2.05  keyword article  transactional      3221.0   
1               LOW  0.05  keyword article  informational      2481.0   
2               LOW  0.00  keyword article  informational      3515.0   
3               LOW  0.00  keyword article     commercial         NaN   
4               LOW  0.00  keyword article  informational      2803.0   

   char_count  ... char_count_tier   ctr  avg_position  engagement_rate  \
0     20457.0  ...     15000-

In [18]:
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
print("Dataset Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

Dataset Shape: (30000, 44)

Columns:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [19]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Total Pages:", len(df))
print("Unique Clients:", df["client_id"].nunique())

print("\nTrend Direction Distribution:")
print(df["trend_direction"].value_counts())

print("\nAverage Content Age (days):")
print(round(df["content_age_days"].mean(), 2))

Total Pages: 30000
Unique Clients: 32

Trend Direction Distribution:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Average Content Age (days):
256.17


The starter dataset contains **30,000 website pages** from **32 clients**. Out of these, **16,262 pages are marked as "down"**, indicating that a large number of pages are experiencing declining performance. The **average content age is 256.17 days**, which suggests that many pages may benefit from regular content updates. These observations support my choice of the **Content Refresh** lane because identifying which pages should be refreshed first can help teams prioritize their work more effectively.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.